mypy/checker.py

```diff
@@ -1901,8 +1901,47 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 rvalue_type = self.expr_checker.accept(rvalue)
                 if not inferred.is_final:
                     rvalue_type = remove_instance_last_known_values(rvalue_type)
+                self.check_forbidden_inference_types(lvalue, rvalue, rvalue_type)
                 self.infer_variable_type(inferred, lvalue, rvalue_type, rvalue)
 
+    def check_forbidden_inference_types(self, lvalue: Lvalue, rvalue: Context, rvalue_type: Type) -> None:
+        if (isinstance(rvalue_type, Instance) and isinstance(lvalue, RefExpr)):
+            if isinstance(rvalue, ListExpr):
+                list_arg_type = rvalue_type.args[0]
+                if isinstance(list_arg_type, Instance):
+                    if list_arg_type.type.fullname() != 'builtins.object':
+                        # This list has a specific type, no further checks needed.
+                        return
+
+                    if self.iterable_has_object_type(iter(rvalue.items)):
+                        return
+
+                    self.msg.need_annotation_for_var(lvalue.node, rvalue)
+
+            elif isinstance(rvalue, DictExpr):
+                dict_value_type = rvalue_type.args[1]
+                if isinstance(dict_value_type, Instance):
+                    if dict_value_type.type.fullname() != 'builtins.object':
+                        # The values in this dict have a specific type, no further checks
+                        # are needed
+                        return
+
+                    if self.iterable_has_object_type(map(lambda x: x[1], rvalue.items)):
```

Comment:
I'm pretty sure this is not optimal, but I couldn't think how to do this a better way

Reply:
You can just pass a list here `[v for (k, v) in rvalue.items]`.